### PR TITLE
Initial implementation of DB

### PR DESCRIPTION
**Pull request by judovana**
This PR is adding xmlrpc request object cache.
Now, before xmlrpc is invoked, cache s consulted. If object is found here, it is used instead of minutes long xmlrpc cache
Cache have is configured via ~/kojiscmplugin-xmlrpc.caching config file.
This file allows to set timeout, since which cache item is invalidated. it allows even to set different times per mehtod, even different times per method@host. It allows to disbale cache (refres hrate to 0) or even disable individual methods, or clean the cache.
The invalidation of cache items is *lazy*.  It means, first thread(s) which request object, which is nolonger valid, will get null, and will thus do full xmlrpc request, and then supply the cache with new value. However any thread(s) requesting  already invalidated object, will still get old value, until it is transparently replaced by firts thread(s).
IMHO  this is better then having synchronised access or then let all the threds to repeat the same request, because each of them got invlaidated item.

### CODE REVIEW

**Comment by judovana** [comment]
Exemplar ~/kojiscmplugin-xmlrpc.caching
```
# copypasted defaults 10 minutes and 6 hours
configRefreshRateMinutes=10
cacheRefreshRateMinutes=360

# hydra is quick enough, any caching on it seems contraproductive (see lower)
blackListedUrlsList=.*hydra.*

# list builds is listing new builds on koji/brew/old api hydra; thus is refrehed once per hour
listBuilds=60

# retaging is touching us very few but unluckily is maybe most used method, and shoudl be cached as much as possible
# set to 24h now, but would be nice to have it longer, as we are testing gating.*, candidate.* and moreover all
# if the build is disbaled, it is usually disabled much later after testing
# on hydra, this method caching should be disabled as we use tag to swap from should be built -> built state on old api
listTags=1440
#listTags@hydra.brq.redhat.com=0

# listrpms and listarchives is listing RPMS/archives of given build.
# although on koji/brew this is immutable, it is not so it on hydra
# Although we can set different timeout of method on different hosts, hydra is currently disabled
# on hydra, those two method caching should be disabled, as we are uploading files in runtime, and can happen, that not fully uplaoded build is picked up
# note, this method should be invoked oly for final, selevted NVR, thus should notbe performance blocker
# note, 99% requests to hydra is via powerfull single qestion new api: getBuildList and getBuildDetail
listRPMs=1440
listArchives=1440
#listRPMs@hydra.brq.redhat.com=0
#listArchives@hydra.brq.redhat.com=0

# this is immutabel onbrew/koji, refgreshing once per week
# I'm not sure how it is on hydra, whee it is hash of File instance. 
getPackageID=10080
#getPackageID@hydra.brq.redhat.com=1440
```

**Comment by judovana** [comment]
Considering this PR as done. Looking forward for feedback

**Comment by judovana** [comment]
cacheRefreshRateMinutes=0

means disabled caching

**Comment by judovana** [comment]
...
listBuilds=30? 10?
...

**Comment by judovana** [comment]
```
cacheReleaseTimeout=10
```
?

**Comment by judovana** [comment]
Exemplar ~/kojiscmplugin-xmlrpc.caching
```
# copypasted defaults 10 minutes and 6 hours
configRefreshRateMinutes=10
cacheRefreshRateMinutes=360
# hopefully there is a lot of ram on hydra
cacheReleaseTimeout=10

# although  hydra is quick enough, it keeps reading FS, lower this  rate at least a bit
#blackListedUrlsList=.*hydra.*

# list builds is listing new builds on koji/brew/old api hydra; thus is refrehed  every 20 minutes
listBuilds=20

# retaging is touching us very few but unluckily is maybe most used method, and shoudl be cached as much as possible
# set to 24h now, but would be nice to have it longer, as we are testing gating.*, candidate.* and moreover all
# if the build is disbaled, it is usually disabled much later after testing
# on hydra, this method caching should be quite small,  as we use tag to swap from should be built -> built state on old api
listTags=1440
listTags@hydra.brq.redhat.com=20

# listrpms and listarchives is listing RPMS/archives of given build.
# although on koji/brew this is immutable, it is not so it on hydra
# on hydra, those two method caching should be disabled, as we are uploading files in runtime, and can happen, that not fully uplaoded build is picked up
# note, this method should be invoked oly for final, selevted NVR, thus should notbe performance blocker
# note, 99% requests to hydra is via powerfull single qestion new api: getBuildList and getBuildDetail
listRPMs=1440
listArchives=1440
listRPMs@hydra.brq.redhat.com=0
listArchives@hydra.brq.redhat.com=0

# this is immutable onbrew/koji, refreshing once per week
# I'm not sure how it is on hydra, where it is hash of File instance. 
getPackageID=10080
getPackageID@hydra.brq.redhat.com=1440

# new api on hydra
getBuildList=20
getBuildDetail=1440
```

Thoughts?

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     * time in minutes, after which the cache records are invlaidate, unless thirs method says differently
+     */
+    public static final String CACHE_REFRESH_RATE_MINUTES = "cacheRefreshRateMinutes";
+
+    /**
+     * N. After acheRefreshRateMinutes 9or per method) x  this number, the item will be removed fromo cache for ever.
+     * The check runs when configRefreshRateMinutes refresh thr config
+     * If set to 0, items remains in cache forever
+     */
+    public static final String CACHE_RELEASE_TIMOUT_MULTIPLIER = "cacheReleaseTimeout";
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java:27]: TIMEOUT

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * space separated list of regexex of  urls
+     * matching urls are NOT cached
+     */
+    public static final String BLACK_LISTED_URLS_LIST = "blackListedUrlsList";
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java:32]: URL_BLACKLIST, urlBlacklist

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+        private boolean alive = true;
+
+        @Override
+        public void run() {
+            while (alive) {
+                try {
+                    Thread.sleep(getConfigRefreshRateMilis());
+                } catch (Exception e) {
+                    LOG.warn("Faield to read existing  cache config", e);
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java:71]: Failed

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+                }
+            }
+            apply();
+            freeOldItems();
+        }
+    }
+
+    protected long toUnits(long time) {
+        return time * minutesToMilis;
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java:93]: Millis

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        }
+
+        private void read() {
+            if (config != null && config.exists()) {
+                try (InputStream inputStream = new FileInputStream(config)) {
+                    Properties propNew = new Properties();
+                    propNew.load(inputStream);
+                    propRaw = propNew;
+                } catch (IOException e) {
+                    LOG.warn("Faield to read existing  cache config", e);
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java:84]: Failed

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/SingleUrlResponseCache.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+        public Date getDateCreated() {
+            return dateCreated;
+        }
+
+        public Object getResult() {
+            return result;
+        }
+
+        public boolean isNotBeingRepalced() {
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/SingleUrlResponseCache.java:57]: Replaced

File: jenkins-scm-koji-plugin/src/main/java/hudson/plugins/scm/koji/client/BuildMatcher.java [hunk 7]
```diff
@@ -117,8 +116,15 @@ public static int compare(Build b1, Build b2) {
         throw new RuntimeException("Unknown order");
     }
 
+    private static final RemoteRequestsCache cache = new RemoteRequestsCache(new File(System.getProperty("user.home"), "kojiscmplugin-xmlrpc.caching"), new OriginalObjectProvider() {
```
Comment (oklinov) [inline jenkins-scm-koji-plugin/src/main/java/hudson/plugins/scm/koji/client/BuildMatcher.java:119]: Too long

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+            cache.put(u.toExternalForm(), sux);
+        }
+        return sux;
+    }
+
+    public void put(final Object result, final URL u, XmlRpcRequestParams params) {
+        ensure(u).put(result, params);
+    }
+
+    public Object get(final URL u, XmlRpcRequestParams params) {
```
Comment (oklinov) [inline koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java:202]: should be private....and Optional :)

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java (hunk 2 above)
Comment (judovana) [reply koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java:32]: nope:P

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java (hunk 1 above)
Comment (judovana) [reply koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestCacheConfigKeys.java:27]: fixed

File: koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java (hunk 3 above)
Comment (judovana) [reply koji-scm-lib/src/main/java/org/fakekoji/xmlrpc/server/expensiveobjectscache/RemoteRequestsCache.java:71]: fixed

**Pull request merged**